packages/mui-material/src/Accordion/Accordion.js

```diff
@@ -46,7 +47,7 @@ const AccordionRoot = styled(Paper, {
     ];
   },
 })(
-  ({ theme }) => {
+  memoTheme(({ theme }) => {
     const transition = {
       duration: theme.transitions.duration.shortest,
```

Comment:
Are we sure about the DX of this? As a developer looking at the source:https://github.com/romgrk/material-ui/blob/8921455731c32f13dd9d61fa49002feb21b673fa/packages/mui-material/src/Accordion/Accordion.js#L36, the first thing I ask myself: why is this verbose like this? Could this be a flag we set?

```diff
 const AccordionRoot = styled(Paper, {
   name: 'MuiAccordion',
   slot: 'Root',
+  onlyTheme: true,
   overridesResolver: (props, styles) => {
     const { ownerState } = props;

     return [
       { [`& .${accordionClasses.region}`]: styles.region },
       styles.root,
       !ownerState.square && styles.rounded,
       !ownerState.disableGutters && styles.gutters,
     ];
   },
 })(
  ({ theme }) => {
```

Reply:
It could have been a flag, I didn't think about it, but I didn't pay too much attention because it can be refactored anytime and I want to iterate on this further.

Emotion is serializing+hashing those style objects on every render of every component, even if we return the same styling object. For internal components we can know, either with the flag on the styled option or a flag on the `memoTheme` return value (e.g. make it return a function with a `.memoized = true` flag), that all the styling only needs to react to `theme` changes. This means we could use emotion's `serializeStyles` when we memoize the style value, and return the serialized object directly, which emotion will handle [here](https://github.com/emotion-js/emotion/blob/0ce3ed0ff3f7d1ddf91db2859f863f230b063e13/packages/serialize/src/index.ts#L211-L238) and skip the `createStringFromObject` on every render/component.

The only thing I need to figure out is variants, we probably need to transfer them from our style value to emotion's serialized style object.

One thing though, the flag would need to affect typings, I hope it's not too complex to setup.

Reply:
I remember why I picked a function though, initially I also had `memoProps` that would memo the value based on which props were accessed (in addition to `memoTheme`), so it was possible to have for the same component style functions that accessed the props and some that didn't (like the [Grid](https://github.com/mui/material-ui/blob/next/packages/mui-material/src/Grid/Grid.js)).

I didn't end up including `memoProps` because the implementation was a bit complex (used a proxy to cache which props values were accessed, like SignalJS props work), and most components didn't access the props so the performance gains should be minimal, though I didn't benchmark the difference. That could improve `Grid` though.

Reply:
> This means we could use emotion's serializeStyles when we memoize the style value, and return the serialized object directly, which emotion will handle [here](https://github.com/emotion-js/emotion/blob/0ce3ed0ff3f7d1ddf91db2859f863f230b063e13/packages/serialize/src/index.ts#L211-L238) and skip the createStringFromObject on every render/component.

Oh, nice, like it used to be in Material UI v4 with JSS. Styles created once per component type.

> The only thing I need to figure out is variants, we probably need to transfer them from our style value to emotion's serialized style object.

Maybe we can create custom class name for each variant and apply them conditionally, like Pigment CSS is doing or like we were doing in Material UI v4. The difference now is that we CSS variables, we have fewer variants to create, e.g. one for color vs. one for color primary, color secondary, etc.

Reply:
Follow-up here: https://github.com/mui/material-ui/pull/43412